server/index.ts

```diff
@@ -347,110 +370,67 @@ app.get("/*", async (req, res, ...args) => {
-    // The URL like me something like
-    // /en-US/docs/HTML/Global_attributes/index.json
-    // and that won't be found in getRedirectUrl() since that doesn't
-    // index things with the '/index.json' suffix. So we need to
-    // temporarily remove it and remember to but it back when we're done.
-    isDocument = true;
-    extraSuffix = "/index.json";
-    lookupURL = lookupURL.replace(extraSuffix, "");
-  } else if (req.path.endsWith("metadata.json")) {
-    isMetadata = true;
-    extraSuffix = "/metadata.json";
-    lookupURL = lookupURL.replace(extraSuffix, "");
-  }
-
-  const isJSONRequest = extraSuffix.endsWith(".json");
-
-  let document;
-  try {
-    console.time(`buildDocumentFromURL(${lookupURL})`);
-    const built = await buildDocumentFromURL(lookupURL);
-    if (built) {
-      document = built.doc;
-    } else if (
-      lookupURL.split("/")[1] &&
-      lookupURL.split("/")[1].toLowerCase() !== DEFAULT_LOCALE.toLowerCase() &&
-      !CONTENT_TRANSLATED_ROOT
-    ) {
-      // Such a common mistake. You try to view a URL that is not en-US but
-      // you forgot to set CONTENT_TRANSLATED_ROOT.
-      console.warn(
-        `URL is for locale '${
-          lookupURL.split("/")[1]
-        }' but CONTENT_TRANSLATED_ROOT is not set. URL will 404.`
-      );
+  });
+  app.get("/[^/]+/docs/*/index.json", async (req, res) => {
+    const url = decodeURI(req.path.replace(/\/index.json$/, ""));
+    try {
+      const doc = await buildDocumentFromURL(url);
+      if (!doc) {
+        return redirectOr404(res, url, "/index.json");
+      }
+      return res.json(doc);
+    } catch (error) {
+      return res.status(500).json(JSON.stringify(error.toString()));
     }
-  } catch (error) {
-    console.error(`Error in buildDocumentFromURL(${lookupURL})`, error);
-    return res.status(500).send(error.toString());
-  } finally {
-    console.timeEnd(`buildDocumentFromURL(${lookupURL})`);
-  }
-
-  if (!document) {
-    const redirectURL = Redirect.resolve(lookupURL);
-    if (redirectURL !== lookupURL) {
-      return res.redirect(301, redirectURL + extraSuffix);
+  });
+  app.get("/[^/]+/docs/*/metadata.json", async (req, res) => {
+    const url = decodeURI(req.path.replace(/\/metadata.json$/, ""));
+    const doc = await buildDocumentFromURL(url);
+    if (!doc) {
+      return redirectOr404(res, url, "/metadata.json");
     }
-    return res
-      .status(404)
-      .sendFile(path.join(STATIC_ROOT, "en-us", "_spas", "404.html"));
-  }
-
-  if (isDocument) {
-    res.json({ doc: document });
-  } else if (isMetadata) {
-    const docString = JSON.stringify({ doc: document });
+    const docString = JSON.stringify(doc);
 
     const hash = crypto.createHash("sha256").update(docString).digest("hex");
-    const { body: _, toc: __, sidebarHTML: ___, ...builtMetadata } = document;
+    const {
+      body: _,
+      toc: __,
+      sidebarHTML: ___,
+      ...builtMetadata
+    } = (doc.doc as any) || {};
```

Comment:
Best to not, do we need the `|| {}`? `doc.doc` should be defined according to the types, and if it's not, we might as well blow everything up:

```suggestion
    } = (doc.doc);
```

Reply:
👍 fixed